### ISSUE TITLE
[no-array-method-this-argument] False positive on non array object but has `map` method

### ISSUE THREAD

**Issue by JounQin** [sindresorhus/eslint-plugin-unicorn#1394]
```ts
export interface ErrorMapperInterface {
  /**
   * Maps an error to a string.
   * @param {string} key the error key. i.e., required, etc
   * @param error the error object
   * @param control control Object (AbstractControl for Angular, NgModel for AngularJS)
   * @returns {string}
   */
  map(key: string, error?: any, control?: any): string;
}

declare const mapper: ErrorMapperInterface

mapper.map('', {})
```

**Comment by fisker** [comment]
I don't think this is the real case, when the first argument is literal, it's ignored, is it a variable?

**Comment by JounQin** [comment]
@fisker Sorry, updated, I just want the example simpler.

**Comment by fisker** [comment]
I don't there is anything we can do, you'll have to use `eslint-disable` https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1385#issuecomment-871028613

**Comment by JounQin** [comment]
When `@typescript-eslint/eslint-parser` is used, we should know whether `mapperService` is a true array.

**Comment by fisker** [comment]
Sure.

**Comment by fregante** [comment]
Twinning this with

- https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1193

**Comment by Setitch** [comment]
I found another false positive, `find` is not Array.find, and the second attribute is current user permissions

```
@Controller('/search')
export class SearchController {
  constructor(
    private readonly service: SearchService,
  ) {
  }
  
  @Get('/home-stream')
  async searchHomeStream(@Query() dto: SearchHomeStreamDto, @Invoker() invoker: UserAuthorization) {
    const data = await this.service.find(dto, invoker);
    /* .... */
  }
}
```


```
@Injectable()
export class SearchService {
  async find(dto: SearchHomeStreamDto, invoker: UserAuthorization): Promise<any[]> { /*....*/ }

 /* ... */
}
```

**Comment by giladgray** [comment]
I am also experiencing this. I have a library called `iterables` that exposes functions with the same names as array methods, but they all take two args. this rule flags them every time, but they are actually valid.

`iterables.find(collection, item => callback)`
